File: stitch.cpp
#include <opencv2/opencv.hpp>
#include "utility.hpp"
#include <opencv2/cudawarping.hpp>
#include <argparse/argparse.hpp>
#include <glob.h>

/**
 * @brief Regular expression of mask image file names.
 * @param n Camera name.
 */
#define MASK_REG_EXP(n) n + ".png"

/**
 * @brief Regular expression of source video file names.
 * @param n Camera name.
 */
#define VID_REG_EXP(n) n + ".mp4"

namespace cuda = cv::cuda;
namespace fs = std::filesystem;

/**
 * @brief Add margins or remove paddings to fit stitched images.
 * @param pjs Projection matrices. This value will be updated.
 * @return Stitched image size.
 */
cv::Size2i crop(std::vector<cv::Mat1d> pjs) {
  std::vector stitched_ltrb = {cv::Point2d(INFINITY, INFINITY), cv::Point2d(-INFINITY, -INFINITY)};
  for (const auto p : pjs) {
    std::vector<cv::Point2d> tf_corners;
    cv::perspectiveTransform((std::vector<cv::Point2d>) {cv::Point2i(0, 0), cv::Point2i(1920, 0), cv::Point2i(0, 1080), cv::Point2i(1920, 1080)}, tf_corners, p);
    stitched_ltrb[0].x = std::min({stitched_ltrb[0].x, tf_corners[0].x, tf_corners[2].x});
    stitched_ltrb[0].y = std::min({stitched_ltrb[0].y, tf_corners[0].y, tf_corners[1].y});
    stitched_ltrb[1].x = std::max({stitched_ltrb[1].x, tf_corners[1].x, tf_corners[3].x});
    stitched_ltrb[1].y = std::max({stitched_ltrb[1].y, tf_corners[2].y, tf_corners[3].y});
  }
  for (auto p : pjs) {
    p = (cv::Mat_<double>(3, 3) <<
      1, 0, -stitched_ltrb[0].x,
      0, 1, -stitched_ltrb[0].y,
      0, 0, 1
    ) * p;
  }

  return cv::Size2i(stitched_ltrb[1].x - stitched_ltrb[0].x, stitched_ltrb[1].y - stitched_ltrb[0].y);
}

cuda::GpuMat stitch(const std::vector<cuda::GpuMat> imgs, const std::vector<cv::Mat1d> pjs, const std::vector<cuda::GpuMat> warped_masks) {
  cuda::GpuMat stitched_img(warped_masks[0].size(), CV_8UC3, cv::Scalar(0));

  for (int i = 0; i < imgs.size(); i++) {
    cuda::GpuMat warped_img;
    cuda::warpPerspective(imgs[i], warped_img, pjs[i], warped_masks[i].size());
    warped_img.copyTo(stitched_img, warped_masks[i]);
  }

  return stitched_img;
}

int main(int argc, char **argv) {
  // parse arguments
  argparse::ArgumentParser parser;
  parser.add_argument("-m", "--mask_dir").required().help("specify mask image directory").metavar("PATH_TO_MASK_DIR");
  parser.add_argument("-p", "--pj_file").required().help("specify projection matrix file").metavar("PATH_TO_PJ_FILE");
  parser.add_argument("-s", "--src_dir").required().help("specify source undistorted video directory").metavar("PATH_TO_SRC_DIR");
  parser.add_argument("-t", "--tgt_file").required().help("specify target video file").metavar("PATH_TO_TGT_FILE");
  parser.parse_args(argc, argv);

  // read projection matrix file
  const auto pj_dict = read_json(parser.get("--pj_file"));

  // load constants
  std::vector<cv::VideoCapture> caps;
  std::vector<cuda::GpuMat> masks;
  std::vector<cv::Mat1d> pjs;
  for (const auto [n, p] : pj_dict.items()) {
    glob_t mask_files, vid_files;
    glob((fs::path(parser.get("--mask_dir")) / (MASK_REG_EXP(n))).c_str(), 0, NULL, &mask_files);
    glob((fs::path(parser.get("--src_dir")) / (VID_REG_EXP(n))).c_str(), 0, NULL, &vid_files);
    if (mask_files.gl_pathc == vid_files.gl_pathc == 1) {
      caps.emplace_back(vid_files.gl_pathv[0]);
      cuda::GpuMat mask;
      mask.upload(cv::imread(mask_files.gl_pathv[0], cv::IMREAD_GRAYSCALE));
      masks.push_back(mask);
      pjs.push_back((cv::Mat_<double>(3, 3) <<
        p["projective_matrix"][0][0], p["projective_matrix"][0][1], p["projective_matrix"][0][2],
        p["projective_matrix"][1][0], p["projective_matrix"][1][1], p["projective_matrix"][1][2],
        p["projective_matrix"][2][0], p["projective_matrix"][2][1], p["projective_matrix"][2][2]
      ));
    }
    globfree(&mask_files);
    globfree(&vid_files);
  }
  if (caps.size() == 0) {
    std::cout << "no mask image file or source video file was found" << std::endl;
    exit(EXIT_FAILURE);
  }

  // prepare constants
  const auto frm_size = crop(pjs);

  fs::create_directories(fs::absolute(parser.get("--tgt_file")).parent_path());
  cv::VideoWriter rec(parser.get("--tgt_file"), cv::VideoWriter::fourcc('m', 'p', '4', 'v'), caps[0].get(cv::CAP_PROP_FPS), frm_size);

  std::vector<cuda::GpuMat> warped_masks;
  for (int i = 0; i < caps.size(); i++) {
      cuda::GpuMat warped_mask;
      cuda::warpPerspective(masks[i], warped_mask, pjs[i], frm_size);
      warped_masks.push_back(warped_mask);
  }

  // stitch
  while (true) {
    std::vector<cuda::GpuMat> frms;
    auto is_eof = false;
    for (auto c : caps) {
      cv::Mat3b frm;
      c >> frm;
      if (frm.empty()) {
        is_eof = true;
        break;
      }
      cuda::GpuMat frm_on_gpu;
      frm_on_gpu.upload(frm);
      frms.push_back(frm_on_gpu);
    }
    if (is_eof) break;

    const auto stitched_frm_on_gpu = stitch(frms, pjs, warped_masks);
    cv::Mat3b stitched_frm;
    stitched_frm_on_gpu.download(stitched_frm);

    rec << stitched_frm;

    if ((int) caps[0].get(cv::CAP_PROP_POS_FRAMES) % 100 == 0) {
      std::cout << (int) caps[0].get(cv::CAP_PROP_POS_FRAMES) << " frames have been completed" << std::endl;
    }
  }

  for (auto c : caps) c.release();
  rec.release();

  return 0;
}

File: stitch.py
import json
import os.path as path
import pickle
from datetime import datetime
from glob import glob
from os import makedirs
import cv2
import numpy as np
from tqdm import tqdm

"""
BEGIN : str
    Begin time to stitch video.
END : str
    End time to stitch video.
MASK_REG_EXP : (n: str) -> str
    Regular expression of mask image file name.
"""

BEGIN = "00:00:00"
END = "23:59:59"
MASK_REG_EXP = lambda n: n + ".png"

def _crop(pjs: dict[str, np.ndarray]) -> tuple[dict[str, np.ndarray], tuple[int, int]]:
    """
    Add margins or remove paddings to fit stitched images.

    Parameters
    ----------
    pjs : dict[str, ndarray[float64]]
        Dictionary of camera names and original projection matrices.

    Returns
    -------
    pjs : dict[str, ndarray[float64]]
        Dictionary of camera names and cropped projection matrices.
    img_size : tuple[int, int]
        Stitched image size.
    """

    stitched_ltrb = [np.inf, np.inf, -np.inf, -np.inf]
    for p in pjs.values():
        tf_corners = cv2.perspectiveTransform(np.array(((0, 0), (1920, 0), (0, 1080), (1920, 1080)), dtype=np.float32)[np.newaxis], p).squeeze(axis=0)
        stitched_ltrb[0] = min(stitched_ltrb[0], tf_corners[0, 0], tf_corners[2, 0])
        stitched_ltrb[1] = min(stitched_ltrb[1], tf_corners[0, 1], tf_corners[1, 1])
        stitched_ltrb[2] = max(stitched_ltrb[2], tf_corners[1, 0], tf_corners[3, 0])
        stitched_ltrb[3] = max(stitched_ltrb[3], tf_corners[2, 1], tf_corners[3, 1])
    pjs = pjs.copy()
    for n, p in pjs.items():
        pjs[n] = np.dot(np.array((
            (1, 0, -stitched_ltrb[0]),
            (0, 1, -stitched_ltrb[1]),
            (0, 0, 1)
        ), dtype=np.float64), p)

    return pjs, (int(stitched_ltrb[2] - stitched_ltrb[0]), int(stitched_ltrb[3] - stitched_ltrb[1]))

def stitch(mask_dir: str, pj_file: str, src_dir: str, tgt_file: str, ts_cache_file: str) -> None:
    # load constants
    begin_in_sec = int((datetime.strptime(BEGIN, "%H:%M:%S") - datetime(1900, 1, 1)).total_seconds())
    end_in_sec = int((datetime.strptime(END, "%H:%M:%S") - datetime(1900, 1, 1)).total_seconds())
    with open(pj_file) as f:
        pj_dict: dict[str, dict[str, int | list[list[float]]]] = json.load(f)
    with open(ts_cache_file, mode="rb") as f:
        ts_cache: tuple[dict[str, list[tuple[int, int]]], dict[str, int]] = pickle.load(f)

    # prepare constants
    cam_names = pj_dict.keys() & ts_cache[0].keys()
    pjs, frm_size = _crop({n: np.array(pj_dict[n]["projective_matrix"], dtype=np.float64) for n in cam_names})
    makedirs(path.dirname(tgt_file), exist_ok=True)
    rec = cv2.VideoWriter(tgt_file, cv2.VideoWriter_fourcc(*"mp4v"), 5, frm_size)
    warped_masks = {n: cv2.warpPerspective(cv2.imread(path.join(mask_dir, MASK_REG_EXP(n)), flags=cv2.IMREAD_GRAYSCALE), pjs[n], frm_size) for n in cam_names}

    # stitch
    status: dict[str, dict[str, int | np.ndarray | cv2.VideoCapture]] = {}
    for cur_in_sec in tqdm(np.arange(begin_in_sec, end_in_sec, step=0.2), desc="stitching"):
        stitched_frm = np.zeros((frm_size[1], frm_size[0], 3), dtype=np.uint8)
        for n in cam_names:
            vid_idx, frm_idx = ts_cache[0][n][int(5 * (cur_in_sec - ts_cache[1][n]))]

            if n in status.keys() and status[n]["vid_idx"] != vid_idx:
                status[n]["cap"].release()
            if n not in status.keys() or status[n]["vid_idx"] != vid_idx:
                status[n] = {"cap": cv2.VideoCapture(filename=glob(path.join(src_dir, f"camera{n}/video_??-??-??_{vid_idx:02d}.mp4"))[0]), "vid_idx": vid_idx}

            while status[n]["cap"].get(cv2.CAP_PROP_POS_FRAMES) <= frm_idx:
                status[n]["frm"] = status[n]["cap"].read()[1]

            cv2.copyTo(cv2.warpPerspective(status[n]["frm"], pjs[n], frm_size), warped_masks[n], dst=stitched_frm)
        rec.write(stitched_frm)

    for s in status.values():
        s["cap"].release()
    rec.release()

if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("-m", "--mask_dir", required=True, help="specify mask image directory", metavar="PATH_TO_MASK_DIR")
    parser.add_argument("-p", "--pj_file", required=True, help="specify projection matrix file", metavar="PATH_TO_PJ_FILE")
    parser.add_argument("-s", "--src_dir", required=True, help="specify source undistorted video directory", metavar="PATH_TO_SRC_DIR")
    parser.add_argument("-t", "--tgt_file", required=True, help="specify target video file", metavar="PATH_TO_TGT_FILE")
    parser.add_argument("-tc", "--ts_cache_file", required=True, help="specify timestamp cache file", metavar="PATH_TO_TS_CACHE_FILE")
    args = parser.parse_args()

    stitch(args.mask_dir, args.pj_file, args.src_dir, args.tgt_file, args.ts_cache_file)
